### PR TITLE
Feature/multiple pom select and add to list of actions

### PR DESCRIPTION
**Pull request by vyasprateek**
=> allow multi POM maped elements selection when adding Actions
=> Add also API Models to Add Actions from
=> Add all Shared Repo items to Add From Shared Repo along with Drag Drop for Actions and Variables
=> AutoLearned replaced with Highlight button (waiting for Mahesh's push to enable it's working)
=> Context changed now handles the ActionsLibrary over AddAction Panel
=> Drap Drop API Models and API Models Folder as well.

### CODE REVIEW

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/APINavPage.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            mContext = context;
+            mItemTypeRootNode = itemTypeRootNode;
+            GingerHelpProvider.SetHelpString(this, itemTypeName.TrimEnd(new char[] { 's' }));
+
+            xTreeView.TreeTitle = itemTypeName;
+            xTreeView.TreeIcon = itemTypeIcon;
+
+            mContext.PropertyChanged += MContext_PropertyChanged;
+            mContext.Activity.PropertyChanged += Activity_PropertyChanged;
+
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/APINavPage.xaml.cs]: why you track  mContext.Activity.PropertyChanged? i guess all you need is not in Context

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/APINavPage.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        private void CurrentActivity_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        {
+            //xTreeView.Tree.TreeNodesFilterByField = new Tuple<string, string>(nameof(ApplicationPOMModel.TargetApplicationKey) + "." + nameof(ApplicationPOMModel.TargetApplicationKey.ItemName), mContext.BusinessFlow.CurrentActivity.TargetApplication);
+            //xTreeView.Tree.FilterType = UCTreeView.eFilteroperationType.Equals;
+        }
+
+        private void MainTreeView_ItemSelected(object sender, EventArgs e)
+        {
+            //GridLength POMDetailsRegionHeight = new GridLength(400, GridUnitType.Star);
+            //GridLength unloadedPOMDetailsHeight = new GridLength(0);
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/APINavPage.xaml.cs]: please delete all commented out code if not needed

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs [hunk 3]
```diff
@@ -59,10 +59,13 @@ public ActionsLibraryNavPage(Context context)
             InitializeComponent();
             mContext = context;
             mActionsList = mContext.BusinessFlow.CurrentActivity.Acts;
+
+            mContext.PropertyChanged += MContext_PropertyChanged;
+            mContext.Activity.PropertyChanged += Activity_PropertyChanged;
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs]: mContext.Activity.PropertyChanged is needed or you can relay only on Context now?

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs [hunk 4]
```diff
@@ -170,6 +197,9 @@ private ObservableList<Act> GetPlatformsActions(bool ShowAll = false)
                 if (a.IsSelectableAction == false)
                     continue;
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs]: does GetPlatformsActions() is duplication to what we have in AddActionPage?

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs [hunk 5]
```diff
@@ -288,17 +318,19 @@ private void AddAction()
         {
             if (ActionsTabs.SelectedContent != null && ((ucGrid)ActionsTabs.SelectedContent).CurrentItem != null)
             {
-                if (((Act)(((ucGrid)ActionsTabs.SelectedContent).CurrentItem)).AddActionWizardPage != null)
+                Act selectedAction = ((ucGrid)ActionsTabs.SelectedContent).CurrentItem as Act;
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs]: AddAction() is duplication to what we have in AddActionpage?

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/MainAddActionsNavigationPage.xaml.cs [hunk 6]
```diff
@@ -41,11 +45,12 @@ public MainAddActionsNavigationPage(Context context)
             xNavigationBarPnl.Visibility = Visibility.Collapsed;
             xSelectedItemFrame.ContentRendered += NavPnlActionFrame_ContentRendered;
             SetRecordButtonAccessebility();
+            xApplicationModelsPnl.Visibility = Visibility.Collapsed;
         }
 
         private void SetRecordButtonAccessebility()
         {
-            Agent mAgent = AgentHelper.GetDriverAgent(mContext.BusinessFlow.CurrentActivity, mContext.Runner, mContext);
+            mAgent = AgentHelper.GetDriverAgent(mContext.BusinessFlow.CurrentActivity, mContext.Runner, mContext);
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/MainAddActionsNavigationPage.xaml.cs]: why we still have this  mAgent = AgentHelper.GetDriverAgent... Agent now exist on Context

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs [hunk 7]
```diff
@@ -41,30 +89,80 @@ public TreeView1 TreeView
...
+            mContext.Activity.PropertyChanged -= Activity_PropertyChanged;
+            mContext.Activity.PropertyChanged += Activity_PropertyChanged;
+            if (e.PropertyName is nameof(mContext.BusinessFlow) || e.PropertyName is nameof(mContext.Activity))
+            {
+                UpdatePOMTree();
+            }
+        }
+
+        private void Activity_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        {
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs]: Activity_PropertyChanged? needed?

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs [hunk 8]
```diff
@@ -41,30 +89,80 @@ public TreeView1 TreeView
...
+            if (mContext.BusinessFlow.CurrentActivity == null)
+                mContext.BusinessFlow.CurrentActivity = mContext.BusinessFlow.Activities[0];
+
+            xTreeView.Tree.TreeNodesFilterByField = new Tuple<string, string>(nameof(ApplicationPOMModel.TargetApplicationKey) + "." + nameof(ApplicationPOMModel.TargetApplicationKey.ItemName), mContext.BusinessFlow.CurrentActivity.TargetApplication);
+            xTreeView.Tree.FilterType = UCTreeView.eFilteroperationType.Equals;
+            xTreeView.Tree.RefresTreeNodeChildrens(mItemTypeRootNode);
+        }
+
+        private void CurrentActivity_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs]: CurrentActivity_PropertyChanged? needed?

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs [hunk 9]
```diff
@@ -142,10 +240,125 @@ private void SetElementsGridView()
...
+                LocateBy = eLocateBy.POMElement,
+                LocateValue = elementInfo.ParentGuid.ToString() + "_" + elementInfo.Guid.ToString(),
+                ElementValue = "",
+                AddPOMToAction = true,
+                POMGuid = elementInfo.ParentGuid.ToString(),
+                ElementGuid = elementInfo.Guid.ToString(),
+                LearnedElementInfo = elementInfo,
+            };
+
+            switch (elementInfo.ElementTypeEnum)
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs]: you not supposed to set the operation of the action in this page, all need to be done by mPlatform.GetPlatformAction() in generic code, what if tomorrow we will have POM of Java application and we will want to add action from one of it elements? so we will do switch case and check if element is Div? no, this is why the action setup needs to be done by the proper platform (web, java, etc.)

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/SharedRepositoryNavPage.xaml [hunk 10]
```diff
@@ -33,6 +33,25 @@
                 </Style>
             </TabControl.Resources>
 
+            <TabItem x:Name="tbActivitiesGroups" Margin="0,0,0,0">
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/SharedRepositoryNavPage.xaml]: 1- not sure why you created all shared repo tabs again and didn't re-used existing SR page we have (inside Frame), but you already done it so keep it
2- you missing 'x' at begining of x:NAme="" at many controls like "x:Name="tbiActivities"" should be x:Name="xSharedActivitiesTab"

File: Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs [hunk 11]
```diff
@@ -183,6 +187,25 @@ private void listActions_ItemDropped(object sender, EventArgs e)
...
+                    if (droppedItem is RepositoryFolder<ApplicationAPIModel>)
+                    {
+                        apiModelsList = (droppedItem as RepositoryFolder<ApplicationAPIModel>).GetFolderItems();
+                    }
+                    else
+                    {
+                        apiModelsList.Add(droppedItem as ApplicationAPIModel);
+                    }
+
+                    AddApiModelActionWizardPage APIModelWizPage = new AddApiModelActionWizardPage(mContext);
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs]: changing the wizard pages from outside is not the way to do.
pass the apiModelsList in AddApiModelActionWizardPage constractor which will have defualt param for it = null, and inside the constractor if the list is not null so hide first page but all behaivor should be handled inside the wizard and not from outside

File: Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs [hunk 12]
```diff
@@ -183,6 +187,25 @@ private void listActions_ItemDropped(object sender, EventArgs e)
...
+                    APIModelWizPage.AAMList = apiModelsList;
+                    APIModelWizPage.Pages.RemoveAt(0);
+                    WizardWindow wizWindow = new WizardWindow(APIModelWizPage);
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs]: what is the below for?
WizardWindow wizWindow = new WizardWindow(APIModelWizPage);
APIModelWizPage.GetCurrentPage().Page.WizardEvent(new WizardEventArgs(APIModelWizPage, EventType.Active));
wizWindow.Show();

only for showing the Wizard? if yes this is not the way, you should use:
WizardWindow.ShowWizard(new APIModelWizPage());

File: Ginger/GingerCore/Drivers/Mobile/Perfecto/PerfectoDriver.cs [hunk 13]
```diff
@@ -473,6 +477,15 @@ private void PressBtn(ePressKey PressKey)
             mDriver.ExecuteScript("mobile:presskey", params1);
         }
 
+        private void TakeScreenShot()
```
Comment (MeniKadosh1) [inline Ginger/GingerCore/Drivers/Mobile/Perfecto/PerfectoDriver.cs]: perfecto changes are not related...
if it's needed pls create PR to master

File: Ginger/Ginger/ApiModelsFolder/AddApiModelActionWizardPage.cs [hunk 14]
```diff
@@ -73,6 +74,9 @@ public override void Finish()
         {
             foreach (ApplicationAPIModel aamb in AAMList)
             {
+                if (aamb.TargetApplicationKey == null || Convert.ToString(aamb.TargetApplicationKey.ItemName) != mContext.Activity.TargetApplication)
```
[the remark this answers is not in the document]
Comment (vyasprateek) [reply Ginger/Ginger/ApiModelsFolder/AddApiModelActionWizardPage.cs]: this extra filtering is for the case when we Drag-Drop the API Folder from the side panel. Without this it would add all the available API's Action into the Actions List

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs (hunk 3 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs]: Actually, after merging with the latest change, i tried with mContext.Target over APIs and POMs pages but that wasn't working so i did not made further changes over the same branch. I'll update the same and push.

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs (hunk 4 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs]: Yes Meni, it's the copy.

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/APINavPage.xaml.cs (hunk 2 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/APINavPage.xaml.cs]: Done !

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/MainAddActionsNavigationPage.xaml.cs (hunk 6 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/MainAddActionsNavigationPage.xaml.cs]: Updated with ContextProperty change

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs (hunk 7 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs]: Not needed, i created branch with changes before Mahesh's push was merged and thought of updating with ContextChange on separate branch later and just resolved conflicts that day.
Now, updated with ContextChange everywhere.

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs (hunk 8 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs]: Not needed, updated with ContextChanged on latest push.

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs (hunk 9 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs]: Okay Done. Updated WebPlatforms's GetPlatformActionByElementInfo with the same.

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/SharedRepositoryNavPage.xaml (hunk 10 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/SharedRepositoryNavPage.xaml]: Done !

File: Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs (hunk 11 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs]: Done!
Overloaded constructor as:
WizardBase wizard = (WizardBase)Activator.CreateInstance(t, mContext);
inside AddAction() started throwing Exception due to mismatching Wizard's Constructor.

File: Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs (hunk 12 above)
Comment (vyasprateek) [reply Ginger/Ginger/BusinessFlowPages_New/ActionsListViewPage.xaml.cs]: Done.

File: Ginger/GingerCore/Drivers/Mobile/Perfecto/PerfectoDriver.cs (hunk 13 above)
Comment (vyasprateek) [reply Ginger/GingerCore/Drivers/Mobile/Perfecto/PerfectoDriver.cs]: Undone Perfecto changes in latest push.

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/MainAddActionsNavigationPage.xaml.cs [hunk 15]
```diff
@@ -72,15 +75,27 @@ private void NavPnlActionFrame_ContentRendered(object sender, EventArgs e)
...
+                    xNavigationBarPnl.Visibility = Visibility.Visible;
+                    xAddActionsOptionsPnl.Visibility = Visibility.Collapsed;
+                    xApplicationModelsPnl.Visibility = Visibility.Visible;
+                }
+                else
+                {
+                    (sender as Frame).Visibility = Visibility.Collapsed;
+                    xNavigationBarPnl.Visibility = Visibility.Collapsed;
+                    xAddActionsOptionsPnl.Visibility = Visibility.Visible;
+                    xApplicationModelsPnl.Visibility = Visibility.Collapsed;
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/MainAddActionsNavigationPage.xaml.cs]: all of this APIModel View need to be in the dedicated page for Application Models and not here in main page

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs [hunk 16]
```diff
@@ -41,30 +86,74 @@ public TreeView1 TreeView
             get { return xTreeView; }
         }
 
-        public POMNavPage(string itemTypeName, eImageType itemTypeIcon, ITreeViewItem itemTypeRootNode, RoutedEventHandler saveAllHandler = null, RoutedEventHandler addHandler = null, EventHandler treeItemDoubleClickHandler = null)
+        public POMNavPage(Context context, string itemTypeName, eImageType itemTypeIcon, ITreeViewItem itemTypeRootNode, RoutedEventHandler saveAllHandler = null, RoutedEventHandler addHandler = null, EventHandler treeItemDoubleClickHandler = null)
         {
             InitializeComponent();
 
+            mContext = context;
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/POMNavPage.xaml.cs]: also here i think you needed to use SingleItemTreeViewSelectionPage unless i'm missing something, let's talk on it

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/SharedRepositoryNavPage.xaml [hunk 17]
```diff
@@ -33,23 +33,42 @@
                 </Style>
             </TabControl.Resources>
```
Comment (MeniKadosh1) [inline Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/SharedRepositoryNavPage.xaml]: also here i see you created duplication of existing SharedRepositoryPage we have so either remove the other one or use it instead

File: Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs [hunk 18]
```diff
@@ -170,6 +184,9 @@ private ObservableList<Act> GetPlatformsActions(bool ShowAll = false)
                 if (a.IsSelectableAction == false)
                     continue;
 
+                if (mContext.BusinessFlow.CurrentActivity == null)
```
[the remark this answers is not in the document]
Comment (MeniKadosh1) [reply Ginger/Ginger/BusinessFlowPages_New/AddActionMenu/ActionsLibraryNavPage.xaml.cs]: i still see this manipulation of Context value, have you done pushing the PR changes?
